src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java

```diff
@@ -162,22 +159,36 @@ private static void renderTerminal( ClientMonitor monitor, float xMargin, float
 
                 if( redraw )
                 {
-                    ByteBuffer monitorBuffer = GLAllocation.createDirectByteBuffer( width * height * 3 );
-                    for( int y = 0; y < height; y++ )
+                    GL15.glBindBuffer( GL31.GL_TEXTURE_BUFFER, monitor.tboBuffer );
+
+                    if( monitor.tboBufferSize != width * height * 3 )
+                    {
+                        monitor.tboBufferSize = width * height * 3;
+                        GL15.glBufferData( GL31.GL_TEXTURE_BUFFER, monitor.tboBufferSize, GL15.GL_DYNAMIC_DRAW );
+                    }
+
+                    monitor.tboContents = GL15.glMapBuffer( GL31.GL_TEXTURE_BUFFER, GL15.GL_WRITE_ONLY, monitor.tboBufferSize, monitor.tboContents );
+                    ByteBuffer monitorBuffer = monitor.tboContents;
+
+                    // When OptiFine is installed, glMapBuffer returns null sometimes for some reason.
+                    // TODO: Investigate this and fix it properly rather than using this hack.
+                    if ( monitorBuffer != null )
```

Comment:
If the buffer is null, does this mean the monitor entirely fails to update? If so, is it worth falling back to a version which creates a new buffer and calls `glBufferData`?

Reply:
I honestly have no idea. In the testing I've done with OptiFine, it seems to work fine, but we do need to test this thoroughly.

> If so, is it worth falling back to a version which creates a new buffer and calls glBufferData?

No, but that gave me an idea. We can fall back to using `glBufferSubData` instead. It's more expensive than `glMapBuffer`, but there's pretty much no good reason why it should fail.

Reply:
Yeah. I'm somewhat surprised it's not printing anything to the logs. I was fairly sure Minecraft dumped `glGetError` after each frame.

Reply:
It does. Which could potentially mean that `glMapBuffer` isn't erroring, [but as the spec states](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glMapBuffer.xhtml):

> If an error is generated, a `NULL` pointer is returned.

There is no other situation I can find in the spec where it should return `null`. Unless there's an exception for LWJGL's bindings?

Reply:
It's worth noting the existence of the [OpenGL 4.3's debugging features](https://www.khronos.org/opengl/wiki/Debug_Output).

Instead of polling `glGetError`, it lets us hand a callback to the driver, which will give us detailed errors, warnings and informational messages. However, I have no idea how well this would interact with Minecraft's existing code.

I don't see how this would give us errors that `glGetError` doesn't detect though.

Reply:
I _think_ Minecraft 1.15 does register this callback, so I might have a fiddle in porting this across. Though yes, don't know if it'd actually reveal anything.